ui/app/components/ui/tabs/tabs.component.js

```diff
@@ -66,4 +74,24 @@ export default class Tabs extends Component {
       </div>
     )
   }
+
+  /**
+   * Returns the
+   * @param {ReactNode} [child] - the child node
+   * @returns {string} a stable ID for this child
+   * @private
+   */
+  _generateKey (child) {
+    return kebabCase(child?.props.name)
+  }
+
+  /**
+   * Returns the index of the child with the given key
+   * @param {string} key - the child key to search for
+   * @returns {number}
+   * @private
+   */
+  _findChildByKey (key) {
+    return React.Children.toArray(this.props.children).findIndex((c) => this._generateKey(c) === key)
```

Comment:
This seems pretty convoluted. Generating these keys upon each render and each click, all so that we can re-order tabs and preserve the preferred tab? I don't see why we'd want that property. I don't know that we've ever re-ordered tabs.

Reply:
I disagree that it it convoluted.

> Generating these keys upon each render

Generating the keys isn't a particularly expensive operation. Iterating the children also isn't particularly expensive, and this method, ` _findChildByKey`, is only run on mount.

> and each click

How does this get run on each click?

Reply:
> I don't know that we've ever re-ordered tabs.

I think it's a useful property, if we _ever_ add a new tab between two tabs or reorder any tabs ever (that are persisted) it is useful for the default to continue working—it isn't the end of the world if not, but it's definitely a useful additional property.

Reply:
if you think about how you would recreate this component using functional components, the generation of each key could be memoized. The initial state could be handled using an initial state setter function
```js
  const [state, setState] = useState(() => _findChildByKey(props))
```

overall I'm not a huge fan of intercepting children and cloning elements, but I don't think there would be an issue overall with performance with the changes suggested here. I do agree that changing the order of tabs, no matter how rare, should still respect saved preference.

Reply:
> I'm not a huge fan of intercepting children and cloning elements

Can we avoid it here? I'm open to not doing this if there's a cleaner way.

Reply:
It could be avoided by passing in the individual tabs via a render prop instead of as children. Or by storing the active tab as Redux state or React context or something.

I would think a render prop is the better pattern to use here, but it's well beyond the scope of this PR.

Reply:
not without a more significant refactor.

The way that I would handle it is by using `createContext` in the `<Tabs>` component to store the currently active tab. then inside the `<Tab>` eject from rendering if the context is not currently active, the context also provides a `onClick` handler to the `<Tab>` to handle setting the parent context state.

I'm sure there are differences in how I've done them in the past and the exact implementation details here so that'd probably have to be worked out, but using context is a great way to lift state up without having to clone element to pass it back down.

Reply:
render props are another way to go but the API tends to be less... pretty? Though I haven't explored the implementation. Agreed with @Gudahtt that it's probably beyond the scope of this change.

Reply:
> It could be avoided by passing in the individual tabs via a render prop instead of as children

I don't understand how this would help. How would you set the props on each child? The only way we have the index/key is because we're looking at each child—I'm not sure how we'd invert that.

Reply:
>How would you set the props on each child?

Well, it'd be the caller's responsibility to wire the props up correctly. Something along these lines maybe:
```
<Tabs
  render={( activeTab, onClick }) => {
    return (
      <>
        <Tab active={activeTab === 0} onClick={onClick.bind(null, 0} />
        <Tab active={activeTab === 1} onClick={onClick.bind(null, 1} />
      </>
    )
  }}
/>
```

On second thought, React context would definitely be a nicer way to do this :sweat_smile: This is not great. Surely there's a better way to do this with render props than what I mocked up here, but yeah.

Reply:
> The way that I would handle it is by using [context]
— @brad-decker 

> React context would definitely be a nicer way
— @Gudahtt

Is using context tangibly different though? Anything we'd pass via context we could pass directly via `cloneElement` since the tabs are direct descendents.

Reply:
Well... it's different in that we wouldn't need to clone anymore. The `<Tab>` passed in via `children` would be exactly what's rendered.

Reply:
Its preference, to me it is cleaner and uses APIs that are currently recommended and are easily typed with typescript. Cloning elements in typescript just reveals some intrinsic complexity that is not visible in js.

🤷 As I said, I'm fine with the changes as is. This PR didn't introduce the element cloning, and it isn't inherently bad.

Reply:
To give a concrete example of why this pattern makes the elements more difficult to type: https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/components/ui/tabs/tab/tab.component.js#L41

I was unable to mark two props as required, because they're injected by `<Tabs>`. So React would throw the warning each render because they were missing from the original element.

Reply:
Re: the original comment, we had talked about using the key or name directly in the `isActive` conditional, but @whymarrh pointed out that we still need to check on mount to assign a default at least.

If we're going to go with the tab name instead of index, I'd suggest using the tab name directly instead of deriving a key from it and using that. It would be more meaningful for any components using this one as well, as the `onTabClick` function would return a value that has some meaning to the parent component.

Reply:
I've updated this to use the name